11 спринт

AlexOtkidach opened the pull request.

AlexOtkidach committed:
Пользовательский ввод и выполнение запроса к API iTunes Search

AlexOtkidach committed:
Добавил 2 плейсхолдера, установил правильную работу длякнопки Х. Осталось доработать работу плейсхолдеров.

AlexOtkidach committed:
Осталось сделать видимость первого плейсхолдера - не найдено
Видимость второго - ошибка сервера

AlexOtkidach committed:
Исправил ошибки появлением плейсхолдера 1, перенес в ресурсы названия.

AlexOtkidach committed:
Скорректировал 1 плейсхолдер

AlexOtkidach committed:
Добавил проверку наличия интернета и скрытие кнопки Обновить в резкльтате проверки.

DmitrijEgorow commented on app/src/main/AndroidManifest.xml:
```diff
@@ -3,6 +3,7 @@
 
     <!-- Разрешение для доступа к интернету -->
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.INTERNET" tools:ignore="WrongManifestParent" />
```
Зачем tools:ignore?

DmitrijEgorow commented on app/src/main/java/com/example/plm2/Song.kt:
```diff
@@ -0,0 +1,9 @@
+    val artistName: String,
+    val trackTime: String,
+    val trackTimeMillis: Long,
+    val artworkUrl100: String
```
Картинка может не прийти

DmitrijEgorow commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -64,7 +65,70 @@
+                    android:layout_width="312dp"
+                    android:layout_height="88dp"
+                    android:layout_centerInParent="true"
+                    android:layout_marginStart="24dp"
```
Лучше вынести в dimen

AlexOtkidach replied:
Понял, перевел в ресурсы.

DmitrijEgorow commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -64,7 +65,70 @@
+
+                <!-- Второй плейсхолдер -->
+                <ImageView
+                    android:id="@+id/secondPlaceholderImageView"
```
Что такое "второй плецсхолдер"? Непонятно

AlexOtkidach replied:
Внес уточнения в название. Если вопрос был в этом.

DmitrijEgorow commented on app/src/main/res/values/dimens.xml:
```diff
@@ -23,6 +23,8 @@
 
     <dimen name="layout_marginTop_artist">25dp</dimen>
     <dimen name="layout_marginStart">66dp</dimen>
+    <dimen name="topButton">375dp</dimen>
```
Лучше использовать чётные размеры, и даже кратные четырём

AlexOtkidach replied:
Поправил, спасибо за замечание

DmitrijEgorow commented on app/src/main/res/values/strings.xml:
```diff
@@ -33,5 +33,9 @@
+    <string name="artist_and_time">%s • %s</string>
+    <string name="refresh_button">Обновить</string>
+    <string name="not_found">Нечего не нашлось</string>
+    <string name="no_connection">Проблемы со связью\n\nЗагрузка не удалась.Проверьте\nподключение к интернету</string>
```
```suggestion
    <string name="no_connection">Проблемы со связью\n\nЗагрузка не удалась. Проверьте\nподключение к интернету</string>
```

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -1,150 +1,299 @@
+                    val songs = response.body()?.results
+                    if (songs != null) {
+                        for (song in songs) {
+                            println("Track Name: ${song.trackName}, Artist Name: ${song.artistName}")
```
Почему не в лог?

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -1,150 +1,299 @@
+                    }
+                    val tracks: List<Track>? = songs?.map { song ->
+                        Track(
+                            trackName = song.trackName ?: "",
```
Зачем здесь elvis-оператор?

AlexOtkidach replied:
Я понимаю что я мог тут применить конструкцию с if-else, но так код выглядит более компактным на мой взгляд. Или можно еще как-то эту конструкцию выразить, о которой я не подумал/не знаю...?

DmitrijEgorow replied:
Nullable типов здесь вроде нет

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -1,150 +1,299 @@
-        return if (s.isNullOrEmpty()) {
-            View.GONE
+    private fun isNetworkAvailable(connectivityManager: ConnectivityManager): Boolean {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
minSdk = 24 в build.gradle
Зачем эта проверка?

AlexOtkidach replied:
Для того чтобы приложение не запускалось на более ранних версиях, тем самым работало только устройствами поддерживающими функционал моего приложения. В целом я могу это убрать, это же не на что не влияет.

DmitrijEgorow replied:
Диктует поддерживаемые версии Android параметр minSdk в build.gradle, приложение не установится на версии ниже, чем minSdk (сейчас 24)
Build.VERSION_CODES.M – это Android Marshmallow (API 23), поэтому эта проверка с такими настройками явно лишняя
И эта условная конструкция в таком контексте описывается немного по-другому, нежели чем "чтобы приложение не запускалось на более ранних версиях, тем самым работало только устройствами поддерживающими функционал моего приложения"

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -1,150 +1,299 @@
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            val network = connectivityManager.activeNetwork
+            val capabilities = connectivityManager.getNetworkCapabilities(network)
+            return capabilities != null && (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) || capabilities.hasTransport(
```
Лучше как-нибудь по-другому разорвать строки

AlexOtkidach replied:
Сделал код более компактными и избежал явных проверок на null.

DmitrijEgorow commented on app/src/main/java/com/example/plm2/SearchActivity.kt:
```diff
@@ -1,150 +1,299 @@
+            val network = connectivityManager.activeNetwork
+            val capabilities = connectivityManager.getNetworkCapabilities(network)
+            return capabilities != null && (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) || capabilities.hasTransport(
+                NetworkCapabilities.TRANSPORT_CELLULAR
```
Редкость, но ещё может быть Bluetooth

AlexOtkidach replied:
Добавил еще проверок)

DmitrijEgorow commented:
Первый запуск приложения

- [ ] Кнопка обновить в тёмной теме должна иметь белую заливку

https://github.com/AlexOtkidach/plm2/assets/36837067/83fba212-c5e0-4942-8aa5-7f1a84ab5c23

DmitrijEgorow commented:
Хорошо бы отцентрировать
![Screenshot_20231124_085903](https://github.com/AlexOtkidach/plm2/assets/36837067/0522562e-eadd-4f90-94c9-ce5c3f4887de)

DmitrijEgorow commented:
Пользователь привык, что Done скрывает клавиатуру

- [ ] Стрелку назад стоит сделать белой


https://github.com/AlexOtkidach/plm2/assets/36837067/5ae88c98-3ded-4524-886e-104b1ad415bf

DmitrijEgorow commented:
- [ ] Заливка поиска в светлой теме должна быть светло-серой, иначе получается низкий контраст текста относительно фона
- [ ] Крестик не должен быть чёрным согласно дизайну, аналогично с лупой
- [ ] Курсор должен быть синего цвета
- [ ] Нижней черты быть не должно

![Screenshot_20231124_090344](https://github.com/AlexOtkidach/plm2/assets/36837067/d3d06094-36eb-493d-a1d8-5c7f1686b602)
![Screenshot_20231124_090329](https://github.com/AlexOtkidach/plm2/assets/36837067/120de725-1441-4dbd-9a62-31cf61db273d)

AlexOtkidach committed:
Перенес размеры в ресурсы. Откорректировал выравнивание TextView

AlexOtkidach committed:
Изменил цвета ресурсов отвечающих за поиск и иконки, альбомы и данные треков.
Исправил логику появление плейсхолдеров.
Добавил сворачивание клавиатуры после ввода done.
Внес множество мелких правок не влияющих на функционал приложения.

DmitrijEgorow commented:
Switch совсем не виден в тёмной теме
![Screenshot_20231127_153318](https://github.com/AlexOtkidach/plm2/assets/36837067/eeca6881-66c7-4f8b-a2d5-08088ca5c457)

DmitrijEgorow commented:
Наверное, не стоит сохранять плейсхолдер с отсутствием результатов при уходе пользователя с экрана
ОБНОВИТЬ -> Обновить, и по высоте стоит сделать её меньше (и с правильным скруглением углов), как в дизайн-макете

https://github.com/AlexOtkidach/plm2/assets/36837067/4d745719-1a04-4ec8-9448-65a0756df3a5

AlexOtkidach committed:
Исправил анимацию кнопки обновить, внес корректные размеры в ресурсы, исправил цвет включенного свитча темы

AlexOtkidach commented:
> Наверное, не стоит сохранять плейсхолдер с отсутствием результатов при уходе пользователя с экрана ОБНОВИТЬ -> Обновить, и по высоте стоит сделать её меньше (и с правильным скруглением углов), как в дизайн-макете
> 
>  Screen_recording_20231127_153422.mp4

Мне кажется тут логика все таки такая: что приложение держит плейсхолдер в памяти потому, что все еще нет интернета, а так как плейсхолдер проверяет его наличие или отсутствие, то в следствии этого он продолжает напоминать, что интернета все еще нет. Ведь как только интернет появится, ошибка сети пропадет и плейсхолдер появится. В условиях задачи нет на этот счет никой информации :

Плейсхолдер (ошибка сервера)
Если поисковый API запрос завершился неудачно (в ответе от сервера присутствует код ошибки), то на экране «Поиск» отображается изображение-заглушка с сообщением, что на сервере произошла ошибка, и кнопка «Обновить».

DmitrijEgorow commented:
> > Наверное, не стоит сохранять плейсхолдер с отсутствием результатов при уходе пользователя с экрана ОБНОВИТЬ -> Обновить, и по высоте стоит сделать её меньше (и с правильным скруглением углов), как в дизайн-макете
> > Screen_recording_20231127_153422.mp4
> 
> Мне кажется тут логика все таки такая: что приложение держит плейсхолдер в памяти потому, что все еще нет интернета, а так как плейсхолдер проверяет его наличие или отсутствие, то в следствии этого он продолжает напоминать, что интернета все еще нет. Ведь как только интернет появится, ошибка сети пропадет и плейсхолдер появится. В условиях задачи нет на этот счет никой информации :
> 
> Плейсхолдер (ошибка сервера) Если поисковый API запрос завершился неудачно (в ответе от сервера присутствует код ошибки), то на экране «Поиск» отображается изображение-заглушка с сообщением, что на сервере произошла ошибка, и кнопка «Обновить».

Хорошо, теперь понял логику

DmitrijEgorow commented:
С точки зрения визуального соответствия дизайну стало красивее
Было -> Стало
<img width="1039" alt="Screenshot 2023-12-03 at 12 45 28" src="https://github.com/AlexOtkidach/plm2/assets/36837067/05d87e2a-88aa-454a-87f7-8de3812180c4">

Один очень важный момент: нельзя хардкодить размеры элементов, у которых отображается текст. Если у устройства будет стоять увеличение текста, то всё, что не влезет в захардкоженную константами область, просто обрежется. Лучше делать вёрстку максимально адаптивной. Подробнее про вёрстку для таких случаев можно почитать https://support.google.com/accessibility/android/answer/12159181


![Screenshot_20231203_122709](https://github.com/AlexOtkidach/plm2/assets/36837067/06034c27-2798-479c-b0d8-52f3e3ef604e)
В итоге, пространство между текстом и границами кнопки по высоте нужно оставить визуально, как есть сейчас (т.е. не увеличивать охранную зону текста), но при увеличении размера текста кнопка должна соответствующим образом растягиваться по ширине и высоте

DmitrijEgorow commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -57,14 +70,84 @@
+                    android:layout_centerHorizontal="true"
+                    android:layout_marginTop="@dimen/padding200dp"
+                    android:layout_below="@+id/secondPlaceholderImageView"
+                    android:clipToOutline="true"
```
Лучше все важные для такого типа кнопки атрибуты вынести в отдельный стиль, похожая кнопка будет на экране медиатеки

AlexOtkidach committed:
Исправил отступ от TextView согласно макета

AlexOtkidach commented:
> С точки зрения визуального соответствия дизайну стало красивее Было -> Стало <img alt="Screenshot 2023-12-03 at 12 45 28" width="1039" src="https://private-user-images.githubusercontent.com/36837067/287509356-05d87e2a-88aa-454a-87f7-8de3812180c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2MjA0NjksIm5iZiI6MTcwMTYyMDE2OSwicGF0aCI6Ii8zNjgzNzA2Ny8yODc1MDkzNTYtMDVkODdlMmEtODhhYS00NTRhLTg3ZjctOGRlMzgxMjE4MGM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDMlMkZ1cy1lYXN0LT
![2023-12-03_20-44-22](https://github.com/AlexOtkidach/plm2/assets/132559666/ab29e892-ed35-4525-ac63-dec776359d18)
ElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAzVDE2MTYwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRmODUxNjRiYzIyYWZlMDhhMzZjZmM3N2Y1YjAzMDdkNDVhMmMwNDBmYzAzYzg5MTU1NzIxM2VjNDViOTQyNjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Mcmak_7AdFCP-CGARRah6FIMwYAvYiIOgAIAVlCyBaw">
> 
> Один очень важный момент: нельзя хардкодить размеры элементов, у которых отображается текст. Если у устройства будет стоять увеличение текста, то всё, что не влезет в захардкоженную константами область, просто обрежется. Лучше делать вёрстку максимально адаптивной. Подробнее про вёрстку для таких случаев можно почитать https://support.google.com/accessibility/android/answer/12159181
> 
> ![Screenshot_20231203_122709](https://private-user-images.githubusercontent.com/36837067/287509514-06034c27-2798-479c-b0d8-52f3e3ef604e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2MjA0NjksIm5iZiI6MTcwMTYyMDE2OSwicGF0aCI6Ii8zNjgzNzA2Ny8yODc1MDk1MTQtMDYwMzRjMjctMjc5OC00NzljLWIwZDgtNTJmM2UzZWY2MDRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAzVDE2MTYwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyNTFkNmM5ZGY3MDc1ZGQ1ZTA5MThlZmRlMjRlMTk4NjYxYTA2ZGUxZjRkMWIyZDQ5NmIxNDRkNjFjNWVkNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V-6wIga2NspmD876xKIEnepvYig8OEhbvA6Ayt6cAzE) В итоге, пространство между текстом и границами кнопки по высоте нужно оставить визуально, как есть сейчас (т.е. не увеличивать охранную зону текста), но при увеличении размера текста кнопка должна соответствующим образом растягиваться по ширине и высоте
![2023-12-03_20-44-22](https://github.com/AlexOtkidach/plm2/assets/132559666/1ce2bb7a-177e-4dc0-aa7d-b67953670d85)

Я исправил отступ от текста, уменьши и сделал согласно дизайна в фигме. Но кнопка не уменьшается и несмотря на эти изменения, я проконсультировался с наставником. Убрал все лишние поля после текста, все размеры соответствуют дизайну, но визуально кнопка все равно немного отличается.

AlexOtkidach commented:
![Снимок экрана 2023-12-03 214029](https://github.com/AlexOtkidach/plm2/assets/132559666/98843940-3ede-4bad-a830-2703b217f6c9)
![Снимок экрана 2023-12-03 214145](https://github.com/AlexOtkidach/plm2/assets/132559666/04664efc-8e07-4fe4-a243-1ffb626b076f)

Размеры вывел в равные

AlexOtkidach committed:
Исправил отступ от TextView согласно макета, закрыл gone

AlexOtkidach committed:
Внес незначительные изменения.
